File: backend/app/routers/proposals.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import or_
from ..database import get_db
from ..models import Proposal, ProposalSection, Activity, User, UserProfile
from ..schemas import (
    ProposalCreate, ProposalResponse, ProposalUpdate,
    ProposalSectionCreate, ProposalSectionResponse, ProposalSectionUpdate,
    ActivityResponse
)
from ..auth import get_current_active_user
import json

router = APIRouter(prefix="/proposals", tags=["proposals"])

@router.post("/", response_model=ProposalResponse)
def create_proposal(
    proposal: ProposalCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # Create proposal
    db_proposal = Proposal(
        title=proposal.title,
        description=proposal.description,
        organization_id=proposal.organization_id,
        priority=proposal.priority,
        deadline=proposal.deadline,
        estimated_value=proposal.estimated_value,
        tags=json.dumps(proposal.tags),
        created_by=current_user.id
    )
    
    # Add assigned users
    if proposal.assigned_to:
        assigned_users = db.query(User).filter(User.id.in_(proposal.assigned_to)).all()
        db_proposal.assigned_users = assigned_users
    
    db.add(db_proposal)
    db.commit()
    db.refresh(db_proposal)
    
    # Log activity
    activity = Activity(
        proposal_id=db_proposal.id,
        user_id=current_user.id,
        action="created",
        details=f"Created proposal: {proposal.title}"
    )
    db.add(activity)
    db.commit()
    
    return db_proposal

@router.get("/", response_model=List[ProposalResponse])
def list_proposals(
    skip: int = 0,
    limit: int = 100,
    status: Optional[str] = None,
    organization_id: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    query = db.query(Proposal)
    
    if status:
        query = query.filter(Proposal.status == status)
    if organization_id:
        query = query.filter(Proposal.organization_id == organization_id)
    
    proposals = query.offset(skip).limit(limit).all()
    return proposals

@router.get("/{proposal_id}", response_model=ProposalResponse)
def get_proposal(
    proposal_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    proposal = db.query(Proposal).filter(Proposal.id == proposal_id).first()
    if not proposal:
        raise HTTPException(status_code=404, detail="Proposal not found")
    return proposal

@router.put("/{proposal_id}", response_model=ProposalResponse)
def update_proposal(
    proposal_id: int,
    proposal_update: ProposalUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    db_proposal = db.query(Proposal).filter(Proposal.id == proposal_id).first()
    if not db_proposal:
        raise HTTPException(status_code=404, detail="Proposal not found")
    
    update_data = proposal_update.dict(exclude_unset=True)
    if 'tags' in update_data:
        update_data['tags'] = json.dumps(update_data['tags'])
    
    for field, value in update_data.items():
        setattr(db_proposal, field, value)
    
    db.commit()
    db.refresh(db_proposal)
    
    # Log activity if status changed
    if proposal_update.status:
        activity = Activity(
            proposal_id=proposal_id,
            user_id=current_user.id,
            action="status_updated",
            details=f"Status changed to: {proposal_update.status}"
        )
        db.add(activity)
        db.commit()
    
    return db_proposal

# Proposal Sections
@router.post("/{proposal_id}/sections", response_model=ProposalSectionResponse)
def create_proposal_section(
    proposal_id: int,
    section: ProposalSectionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # Verify proposal exists
    proposal = db.query(Proposal).filter(Proposal.id == proposal_id).first()
    if not proposal:
        raise HTTPException(status_code=404, detail="Proposal not found")
    
    db_section = ProposalSection(
        proposal_id=proposal_id,
        title=section.title,
        content=section.content,
        section_type=section.section_type,
        order=section.order,
        last_edited_by=current_user.id
    )
    db.add(db_section)
    db.commit()
    db.refresh(db_section)
    
    # Log activity
    activity = Activity(
        proposal_id=proposal_id,
        user_id=current_user.id,
        action="section_created",
        details=f"Created section: {section.title}"
    )
    db.add(activity)
    db.commit()
    
    return db_section

@router.get("/{proposal_id}/sections", response_model=List[ProposalSectionResponse])
def get_proposal_sections(
    proposal_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    sections = db.query(ProposalSection).filter(
        ProposalSection.proposal_id == proposal_id
    ).order_by(ProposalSection.order).all()
    return sections

@router.put("/{proposal_id}/sections/{section_id}", response_model=ProposalSectionResponse)
def update_proposal_section(
    proposal_id: int,
    section_id: int,
    section_update: ProposalSectionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    db_section = db.query(ProposalSection).filter(
        ProposalSection.id == section_id,
        ProposalSection.proposal_id == proposal_id
    ).first()
    
    if not db_section:
        raise HTTPException(status_code=404, detail="Section not found")
    
    update_data = section_update.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(db_section, field, value)
    
    db_section.last_edited_by = current_user.id
    db_section.version += 1
    
    db.commit()
    db.refresh(db_section)
    return db_section

@router.get("/{proposal_id}/activities", response_model=List[ActivityResponse])
def get_proposal_activities(
    proposal_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    activities = db.query(Activity).filter(
        Activity.proposal_id == proposal_id
    ).order_by(Activity.timestamp.desc()).all()
    return activities

File: convex/proposals.ts
import { query, mutation } from "./_generated/server";
import { v } from "convex/values";
import { getAuthUserId } from "@convex-dev/auth/server";

// Helper function to get authenticated user
async function getAuthenticatedUser(ctx: any) {
  const userId = await getAuthUserId(ctx);
  if (!userId) {
    throw new Error("Not authenticated");
  }
  return userId;
}

// Get user profile with permissions
export const getUserProfile = query({
  args: {},
  handler: async (ctx) => {
    const userId = await getAuthenticatedUser(ctx);
    const profile = await ctx.db
      .query("userProfiles")
      .withIndex("by_user", (q) => q.eq("userId", userId))
      .first();
    
    if (!profile) {
      // Create default profile for new users
      return {
        userId,
        role: "presales" as const,
        department: "Sales",
        permissions: ["read", "write"],
        isActive: true,
      };
    }
    
    return profile;
  },
});

// Create user profile
export const createUserProfile = mutation({
  args: {
    role: v.union(v.literal("admin"), v.literal("manager"), v.literal("presales"), v.literal("viewer")),
    department: v.string(),
  },
  handler: async (ctx, args) => {
    const userId = await getAuthenticatedUser(ctx);
    
    const existing = await ctx.db
      .query("userProfiles")
      .withIndex("by_user", (q) => q.eq("userId", userId))
      .first();
    
    if (existing) {
      throw new Error("User profile already exists");
    }
    
    const permissions = args.role === "admin" 
      ? ["read", "write", "delete", "approve", "manage_users"]
      : args.role === "manager"
      ? ["read", "write", "approve"]
      : ["read", "write"];
    
    return await ctx.db.insert("userProfiles", {
      userId,
      role: args.role,
      department: args.department,
      permissions,
      isActive: true,
    });
  },
});

// List all proposals with filters
export const listProposals = query({
  args: {
    status: v.optional(v.union(
      v.literal("draft"),
      v.literal("in_review"),
      v.literal("approved"),
      v.literal("submitted"),
      v.literal("won"),
      v.literal("lost")
    )),
    organizationId: v.optional(v.id("organizations")),
  },
  handler: async (ctx, args) => {
    const userId = await getAuthenticatedUser(ctx);
    
    let proposals;
    
    if (args.status && args.organizationId) {
      // If both filters are provided, use organization filter first
      proposals = await ctx.db
        .query("proposals")
        .withIndex("by_organization", (q) => q.eq("organizationId", args.organizationId!))
        .filter((q) => q.eq(q.field("status"), args.status!))
        .collect();
    } else if (args.status) {
      proposals = await ctx.db
        .query("proposals")
        .withIndex("by_status", (q) => q.eq("status", args.status!))
        .collect();
    } else if (args.organizationId) {
      proposals = await ctx.db
        .query("proposals")
        .withIndex("by_organization", (q) => q.eq("organizationId", args.organizationId!))
        .collect();
    } else {
      proposals = await ctx.db.query("proposals").collect();
    }
    
    // Get organization details for each proposal
    const proposalsWithOrgs = await Promise.all(
      proposals.map(async (proposal) => {
        const organization = await ctx.db.get(proposal.organizationId);
        const creator = await ctx.db.get(proposal.createdBy);
        return {
          ...proposal,
          organization,
          creator: creator ? { name: creator.name, email: creator.email } : null,
        };
      })
    );
    
    return proposalsWithOrgs;
  },
});

// Get proposal details
export const getProposal = query({
  args: { proposalId: v.id("proposals") },
  handler: async (ctx, args) => {
    const userId = await getAuthenticatedUser(ctx);
    
    const proposal = await ctx.db.get(args.proposalId);
    if (!proposal) {
      throw new Error("Proposal not found");
    }
    
    const organization = await ctx.db.get(proposal.organizationId);
    const creator = await ctx.db.get(proposal.createdBy);
    
    // Get assigned users
    const assignedUsers = await Promise.all(
      proposal.assignedTo.map(async (userId) => {
        const user = await ctx.db.get(userId);
        return user ? { _id: user._id, name: user.name, email: user.email } : null;
      })
    );
    
    return {
      ...proposal,
      organization,
      creator: creator ? { name: creator.name, email: creator.email } : null,
      assignedUsers: assignedUsers.filter(Boolean),
    };
  },
});

// Create new proposal
export const createProposal = mutation({
  args: {
    title: v.string(),
    description: v.string(),
    organizationId: v.id("organizations"),
    priority: v.union(v.literal("low"), v.literal("medium"), v.literal("high"), v.literal("critical")),
    deadline: v.number(),
    estimatedValue: v.number(),
    assignedTo: v.array(v.id("users")),
    tags: v.array(v.string()),
  },
  handler: async (ctx, args) => {
    const userId = await getAuthenticatedUser(ctx);
    
    const proposalId = await ctx.db.insert("proposals", {
      title: args.title,
      description: args.description,
      organizationId: args.organizationId,
      status: "draft",
      priority: args.priority,
      deadline: args.deadline,
      estimatedValue: args.estimatedValue,
      createdBy: userId,
      assignedTo: args.assignedTo,
      tags: args.tags,
      currentVersion: 1,
      isTemplate: false,
    });
    
    // Log activity
    await ctx.db.insert("activities", {
      proposalId,
      userId,
      action: "created",
      details: `Created proposal: ${args.title}`,
      timestamp: Date.now(),
    });
    
    return proposalId;
  },
});

// Update proposal status
export const updateProposalStatus = mutation({
  args: {
    proposalId: v.id("proposals"),
    status: v.union(
      v.literal("draft"),
      v.literal("in_review"),
      v.literal("approved"),
      v.literal("submitted"),
      v.literal("won"),
      v.literal("lost")
    ),
  },
  handler: async (ctx, args) => {
    const userId = await getAuthenticatedUser(ctx);
    
    const proposal = await ctx.db.get(args.proposalId);
    if (!proposal) {
      throw new Error("Proposal not found");
    }
    
    await ctx.db.patch(args.proposalId, {
      status: args.status,
    });
    
    // Log activity
    await ctx.db.insert("activities", {
      proposalId: args.proposalId,
      userId,
      action: "status_updated",
      details: `Status changed to: ${args.status}`,
      timestamp: Date.now(),
    });
    
    return args.proposalId;
  },
});

// Get proposal sections
export const getProposalSections = query({
  args: { proposalId: v.id("proposals") },
  handler: async (ctx, args) => {
    const userId = await getAuthenticatedUser(ctx);
    
    const sections = await ctx.db
      .query("proposalSections")
      .withIndex("by_proposal_order", (q) => q.eq("proposalId", args.proposalId))
      .collect();
    
    // Get editor details for each section
    const sectionsWithEditors = await Promise.all(
      sections.map(async (section) => {
        const editor = await ctx.db.get(section.lastEditedBy);
        return {
          ...section,
          lastEditedBy: editor ? { name: editor.name, email: editor.email } : null,
        };
      })
    );
    
    return sectionsWithEditors.sort((a, b) => a.order - b.order);
  },
});

// Create or update proposal section
export const upsertProposalSection = mutation({
  args: {
    proposalId: v.id("proposals"),
    sectionId: v.optional(v.id("proposalSections")),
    title: v.string(),
    content: v.string(),
    sectionType: v.union(
      v.literal("executive_summary"),
      v.literal("problem_statement"),
      v.literal("solution"),
      v.literal("timeline"),
      v.literal("pricing"),
      v.literal("team"),
      v.literal("case_studies"),
      v.literal("appendix")
    ),
    order: v.number(),
  },
  handler: async (ctx, args) => {
    const userId = await getAuthenticatedUser(ctx);
    
    if (args.sectionId) {
      // Update existing section
      const section = await ctx.db.get(args.sectionId);
      if (!section) {
        throw new Error("Section not found");
      }
      
      await ctx.db.patch(args.sectionId, {
        title: args.title,
        content: args.content,
        sectionType: args.sectionType,
        order: args.order,
        lastEditedBy: userId,
        version: section.version + 1,
      });
      
      return args.sectionId;
    } else {
      // Create new section
      const sectionId = await ctx.db.insert("proposalSections", {
        proposalId: args.proposalId,
        title: args.title,
        content: args.content,
        sectionType: args.sectionType,
        order: args.order,
        lastEditedBy: userId,
        version: 1,
        isLocked: false,
      });
      
      // Log activity
      await ctx.db.insert("activities", {
        proposalId: args.proposalId,
        userId,
        action: "section_created",
        details: `Created section: ${args.title}`,
        timestamp: Date.now(),
      });
      
      return sectionId;
    }
  },
});
